docs/proposals/dramatically-simplify-cluster-creation.md

```diff
@@ -0,0 +1,222 @@
+
+```
+laptop$ kubeadm --help
+
+kubeadm: bootstrap a kubernetes cluster.
+
+Global options:
+
+    --discovery    Discovery mechanism to use. Choose from: gossip,
+                   out-of-band. Default: gossip.
+
+Subcommands:
+    init master    Run this on the first server you deploy onto.
+
+    join node      Run this on other servers to join an existing cluster.
+
+    user create    Create user and fetch authentication config for them.
+```
+
+### Install
+
+*On first machine:*
+
+```
+master# kubeadm init master
+Initializing kubernetes master...  [done]
+Cluster token: 73R2SIPM739TNZOA
+Run the following command on machines you want to become nodes:
+  kubeadm join node --token=73R2SIPM739TNZOA <master-ip>
+You can now run kubectl here.
+```
+
+*On N "node" machines:*
+
+```
+node# kubeadm join node --token=73R2SIPM739TNZOA <master-ip>
+Initializing kubernetes node...    [done]
+Bootstrapping certificates...      [done]
+Joined node to cluster, see 'kubectl get nodes' on master.
+```
+
+Note `[done]` would be colored green in all of the above.
+
+### Install: alternative for automated deploy
+
+*The user (or their config management system) creates a token and passes the same one to both init and join.*
+
+```
+master# kubeadm init master --token=73R2SIPM739TNZOA
+Initializing kubernetes master...  [done]
+You can now run kubectl here.
+```
+
+### Pre-flight check
+
+```
+master# kubeadm init master
+Error: socat not installed. Unable to proceed.
+```
+
+### Control
+
+*On master, after Install, kubectl is automatically able to talk to localhost:8080:*
+
+```
+master# kubectl get pods
+[normal kubectl output]
+```
+
+*To mint new user credentials on the master:*
+
+```
+master# kubeadm user create -o kubeconfig-bob bob
+
+Waiting for cluster to become ready...       [done]
+Creating user certificate for user...        [done]
+Waiting for user certificate to be signed... [done]
+Your cluster configuration file has been saved in kubeconfig.
+
+laptop# scp <master-ip>:/root/kubeconfig-bob ~/.kubeconfig
+laptop# kubectl get pods
+[normal kubectl output]
+```
+
+### Install-addons
+
+*Using CNI network as example:*
+
+```
+master# kubectl apply --purge -f \
+    https://git.io/kubernetes-addons/<X>.yaml
+[normal kubectl apply output]
+```
+
+### Add-node
+
+*Same as Install – "on node machines".*
+
+### Secure
+
+```
+node# kubeadm join --token=GARBAGE node <master-ip>
+Unable to join mesh network. Check your token.
+```
+
+## Work streams – critical path
+
+1. [Debs](https://github.com/kubernetes/release/pull/35) and [RPMs](https://github.com/kubernetes/release/pull/50) (and binaries?) - so that kubernetes can be installed in the first place
+
+    * Mike & Devan
+
+2. [Cluster bootstrap with gossip](https://github.com/kubernetes/kubernetes/pull/30361) - so that we can implement the simple UX with no dependencies
```

Comment:
The [gossip library](https://www.infoq.com/presentations/weave-mesh) we're proposing supports [NaCL based](https://godoc.org/golang.org/x/crypto/nacl/box) encryption based on a shared secret. This short secret can easily be exchanged by the user on the commandline (out of band) in the UX transcripts, and it can then form a backbone for the rest of the bootstrap: getting the cacert (and the address of the API server, which may change later – it can also provide a mechanism for distributing changes to both of these items over time) securely to all joining nodes: see https://github.com/kubernetes/kubernetes/pull/30360/files#diff-ff4da8fc3dcf231fcc5f10f9af427e9cR88